src/plus/webviews/graph/graphWebview.ts

```diff
@@ -706,44 +708,83 @@ export class GraphWebview extends WebviewBase<State> {
 			}
 
 			const metadata = { ...this._refsMetadata.get(id) };
-			if (type !== 'pullRequests') {
+			if (type !== 'pullRequests' && type !== 'upstream') {
 				(metadata as any)[type] = null;
-				this._refsMetadata.set(id, metadata);
+				if (Object.keys(metadata).length === 0) {
+					this._refsMetadata.set(id, metadata);
+				}
+
 				return;
 			}
 
-			const branch = (await this.container.git.getBranches(repoPath, { filter: b => b.id === id && b.remote }))
-				?.values?.[0];
-			const pr = await branch?.getAssociatedPullRequest();
-			if (pr == null) {
-				if (metadata.pullRequests === undefined || metadata.pullRequests?.length === 0) {
-					metadata.pullRequests = null;
+			let branch;
+			if (type === 'pullRequests') {
+				branch = (await this.container.git.getBranches(repoPath, { filter: b => b.id === id && b.remote }))
+					?.values?.[0];
+				const pr = await branch?.getAssociatedPullRequest();
+
+				if (pr == null) {
+					if (metadata.pullRequests === undefined || metadata.pullRequests?.length === 0) {
+						metadata.pullRequests = null;
+					}
+					this._refsMetadata.set(id, metadata);
+					return;
 				}
+
+				const prMetadata: GraphPullRequestMetadata = {
+					// TODO@eamodio: This is iffy, but works right now since `github` and `gitlab` are the only values possible currently
+					hostingServiceType: pr.provider.id as GraphHostingServiceType,
+					id: Number.parseInt(pr.id) || 0,
+					title: pr.title,
+					author: pr.author.name,
+					date: (pr.mergedDate ?? pr.closedDate ?? pr.date)?.getTime(),
+					state: pr.state,
+					url: pr.url,
+					context: serializeWebviewItemContext<GraphItemContext>({
+						webviewItem: 'gitlens:pullrequest',
+						webviewItemValue: {
+							type: 'pullrequest',
+							id: pr.id,
+							url: pr.url,
+						},
+					}),
+				};
+
+				metadata.pullRequests = [prMetadata];
+
 				this._refsMetadata.set(id, metadata);
 				return;
 			}
 
-			const prMetadata: GraphPullRequestMetadata = {
-				// TODO@eamodio: This is iffy, but works right now since `github` and `gitlab` are the only values possible currently
-				hostingServiceType: pr.provider.id as GraphHostingServiceType,
-				id: Number.parseInt(pr.id) || 0,
-				title: pr.title,
-				author: pr.author.name,
-				date: (pr.mergedDate ?? pr.closedDate ?? pr.date)?.getTime(),
-				state: pr.state,
-				url: pr.url,
-				context: serializeWebviewItemContext<GraphItemContext>({
-					webviewItem: 'gitlens:pullrequest',
-					webviewItemValue: {
-						type: 'pullrequest',
-						id: pr.id,
-						url: pr.url,
-					},
-				}),
-			};
+			if (type === 'upstream') {
+				branch = (await this.container.git.getBranches(repoPath, { filter: b => b.id === id && Boolean(b.upstream) }))
```

Comment:
It would be good to re-use the branch from above, if it already existed

Reply:
Two issues with that, assuming you mean the `branch` from the `if` block for `pullRequests` type:
1. The `branch` we get there uses a different filter and is looking for a remote branch rather than a local.
2. We always `return` at the end of that `if` block, so we wouldn't reach here if it was defined there.

Reply:
@axosoft-ramint both cases we'd be looking for the remote branch (we are just doing it slightly differently but should be the same ).

But I missed the return. I was thinking we could get asked to fill out the PR and the branch status at the same time it's not setup like that now. I'll dig more into it later tonight.

Reply:
> @axosoft-ramint both cases we'd be looking for the remote branch (we are just doing it slightly differently but should be the same ).
> 

Oh, really? I thought in the `upstream` case we were looking for a local branch _with_ an upstream remote, not a remote branch? Or is it actually just grabbing the remote one?

> But I missed the return. I was thinking we could get asked to fill out the PR and the branch status at the same time it's not setup like that now. I'll dig more into it later tonight.

Much appreciated!

Reply:
@eamodio Okay, I refactored `getRefMetadata` to get the branch just once for an id, and loop through the missing types internally. Have a look at the latest commit: [Move missing type loop into getRefMetadata, get branch once for id](https://github.com/gitkraken/vscode-gitlens/pull/2441/commits/6751242cbd6a03adbed1c10405c36ff02bcf7976)